### PR TITLE
add mysql-connector-c++

### PR DESCRIPTION
**Pull request by Dentrax**
Fixes:

Related:

#### For new package PRs only

- [ ] This PR is marked as fixing a pre-existing package request bug
  - [ ] Alternatively, the PR is marked as related to a pre-existing package request bug, such as a dependency
- [ ] REQUIRED - The package is available under an OSI-approved or FSF-approved license
- [ ] REQUIRED - The version of the package is still receiving security updates
- [ ] This PR links to the upstream project's support policy (e.g. `endoflife.date`)

#### For new version streams

- [ ] The upstream project actually supports multiple concurrent versions.
- [ ] Any subpackages include the version string in their package name (e.g. `name: ${{package.name}}-compat`)
- [ ] The package (and subpackages) `provides:` logical unversioned forms of the package (e.g. `nodejs`, `nodejs-lts`)
- [ ] If non-streamed package names no longer built, open PR to withdraw them (see [WITHDRAWING PACKAGES](https://github.com/wolfi-dev/os/blob/main/WITHDRAWING_PACKAGES.md))

#### For package updates (renames) in the base images

When updating packages part of base images (i.e. cgr.dev/chainguard/wolfi-base or ghcr.io/wolfi-dev/sdk)
- [ ] REQUIRED cgr.dev/chainguard/wolfi-base and ghcr.io/wolfi-dev/sdk images successfully build
- [ ] REQUIRED cgr.dev/chainguard/wolfi-base and ghcr.io/wolfi-dev/sdk contain no obsolete (no longer built) packages
- [ ] Upon launch, does `apk upgrade --latest` successfully upgrades packages or performs no actions

#### For security-related PRs

- [ ] The security fix is recorded in the [advisories](https://github.com/wolfi-dev/advisories) repo

#### For version bump PRs

- [ ] The `epoch` field is reset to 0

#### For PRs that add patches

- [ ] Patch source is documented

### CODE REVIEW

File: mysql-connector-cpp.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  contents:
+    packages:
+      - autoconf
+      - automake
+      - build-base
+      - busybox
+      - ca-certificates-bundle
+      - cmake
+      - openssl-dev
+      - zlib-dev
```
Comment (xnox) [inline mysql-connector-cpp.yaml:19]: Please use as many system packages as possible. I.e. protobuf zlib lz4 zstd. We do not want vendored copies of those; if at all possible to use system ones.

File: mysql-connector-cpp.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        - ${{package.name}}
+    description: ${{package.name}} dev
+
+  - name: libmysqlclient
+    description: MySQL Connector/C++ shared libraries
+    pipeline:
+      - runs: |
+          mkdir -p ${{targets.contextdir}}/usr/lib
+          mv ${{targets.destdir}}/usr/lib/*.so* ${{targets.contextdir}}/usr/lib/
+          chmod 755 ${{targets.subpkgdir}}/usr/lib/*.so*
```
Comment (xnox) [inline mysql-connector-cpp.yaml:56]: After these splits. The main package contains incorrectly installed cmake files and INFO_BIN and INFO_SRC files. Imho the main package should just have the shared library.

The cmake files should be instructed to be correctly installed into -dev package. Ditto INFO_BIN and INFO_SRC files, not sure how they are used even, if at all.

And imho libmysqlclient as a sub-pckage should not exist at all.

**Review by xnox — CHANGES_REQUESTED** [review]

**Comment by xnox** [comment]
> 2024/10/01 18:44:31 INFO Package: mysql-connector-cpp: [no-repeated-deps]: package zlib-dev is duplicated in environment (ERROR)

File: mysql-connector-cpp.yaml (hunk 1 above)
Comment (Dentrax) [reply mysql-connector-cpp.yaml:23]: Thanks done. Also passed the following flags:
```
-DWITH_BOOST=system \
-DWITH_LZ4=system \
-DWITH_SSL=system \
-DWITH_ZLIB=system \
-DWITH_ZSTD=system
```

But passing `-DWITH_PROTOBUF=system` fails with syntax error due to probuf version, i guess.

```
/home/build/cdk/protocol/mysqlx/protocol.cc:131:17: error: 'log_handler' was not declared in this scope
/home/build/cdk/protocol/mysqlx/protocol.cc:131:3: error: 'SetLogHandler' was not declared in this scope
```

File: mysql-connector-cpp.yaml (hunk 2 above)
Comment (Dentrax) [reply mysql-connector-cpp.yaml:68]: How should we proceed here? I was following the official build doc: https://dev.mysql.com/doc/connector-cpp/9.0/en/connector-cpp-installation-source-cpp.html#connector-cpp-installation-source-build

Do you think `libmysqlclient` package would redundant?

File: mysql-connector-cpp.yaml (hunk 1 above)
Comment (xnox) [reply mysql-connector-cpp.yaml:23]: sadness, likely everything needs to be regenerated or it is incompatible. Leave it as it is, we have tried =)

File: mysql-connector-cpp.yaml (hunk 2 above)
Comment (xnox) [reply mysql-connector-cpp.yaml:68]: the guide only lists libraries, it doesn't tell you the name of the subpackage to use, and how one splits them or not.

So yes, I do believe the libmysqlclient subpackage should be re-dundant, and the default files in destdir should be cleaned up. (correct cmake files locations, and likely purge the weird INFO files if they are unreferenced otherwiese).

Comment (Dentrax) [reply mysql-connector-cpp.yaml]: Thanks @xnox, I have removed the sub-package. But I'm not sure if I understand the "INFO infos", could you please take a look when possible?

Comment (Dentrax) [reply mysql-connector-cpp.yaml]: Kindly ping here. 🙃 @xnox

**Review by EyeCantCU — DISMISSED** [review]

**Review by xnox — APPROVED** [review]

**Pull request merged**